### ISSUE TITLE
Bug: ModuleEvaluation had breaking change to not return completion value

### ISSUE THREAD

**Issue by bmeck** [tc39/ecma262#1227]
Somehow missed this change a while back

This value is used by Node in order to create some reflection https://github.com/nodejs/node/blob/master/lib/internal/modules/esm/create_dynamic_module.js

**Comment by domenic** [comment]
Can you explain the issue in more detail? Remember that the spec has a blanket clause allowing returning JS values which implicitly gives you a normal completion for that value.

**Comment by ljharb** [comment]
Can you link to the part of the spec in question?

**Comment by bmeck** [comment]
@domenic does that apply to concrete methods? I thought that concrete methods needed to return the value the spec says they return.

https://www.ecma-international.org/ecma-262/6.0/index.html#sec-moduleevaluation states that it completes with `result`

newer spec https://tc39.github.io/ecma262/#sec-moduleevaluation completes with `undefined`

**Comment by domenic** [comment]
Yes, it applies to everything. "Concrete methods" are just abstract operations that conform to some interface.

**Comment by bmeck** [comment]
@domenic it still looks like this changed from `result` to `undefined`.

**Comment by domenic** [comment]
Right, that makes sense, since no caller uses that value.

**Comment by bmeck** [comment]
@domenic maybe, I'm not sure on this though since we do have some exposure of the result being used by a host platform. From a spec point of view this seems fine if we don't consider host usage, but I'm pointing particularly that a host is using this result.

**Comment by ljharb** [comment]
This seems more like a layering issue than a spec bug then.

**Comment by littledan** [comment]
It's not clear to me what's going on here exactly. If there are some hosts which concretely point to an abstract algorithm and say, this operation of mine calls that abstract operation and returns the result, it may be helpful for that host's documentation to have stability in the JS spec. This is sort of a layering issue, like for HTML calling into abstract operations.

I'm not sure what abstract operation you're talking about (I don't see one called ModuleEvaluation) and what host API is referring to that. I'm also not sure which spec change made a difference. It would be helpful if you could give these details.

**Comment by bmeck** [comment]
@littledan it used to be called ModuleEvaluation now it is just Evaluate, I have links above. I chose to use the original name prior to some reorganization that went on since that had the behavior desired and is the name of the link to the section still. 

[Evaluate](https://tc39.github.io/ecma262/#sec-moduleevaluation) on Source Text Module Records is the concrete algorithm in question. This corresponds to the API in v8 for [v8::Module::Evaluate](https://github.com/v8/v8/blob/75f14dbffccda0910274c72ffe0cfb0229e6c2b9/include/v8.h#L1325)

**Comment by littledan** [comment]
Ah, sorry, missed that, thanks for the references!
